words/core.asm

```diff
@@ -5309,59 +5220,76 @@ z_s_to_d:       rts
+        ; When we enter, WORKWORD is pointing to the nt_ of this word in the
         ; Dictionary, DP to the previous word, and CP to the next free byte.
+        ; See more details in create_common which sets the stage for us.
+        ;
         ; A Forth definition would be (see "Starting Forth"):
         ; : POSTPONE EXIT  REVEAL POSTPONE ; [ ; IMMEDIATE  Following the
         ; practice of Gforth, we warn here if a word has been redefined.
         ; """
 
 xt_semicolon:
 w_semicolon:
+                dex
+                dex
+                lda workword
+                sta 0,x
+                lda workword+1
+                sta 1,x                 ; ( xt|nt )
+
                 ; Check if this is a : word or a :NONAME word.
                 bit status              ; check bit 6 (overflow flag)
                 bvs _colonword
 
                 ; This is a :NONAME word - just put an RTS on the end and
-                ; the address (held in workword) on the stack.
+                ; leave workword (xt) on the stack.
                 lda #OpRTS
                 jsr cmpl_a
 
-                dex
-                dex
-                lda workword
-                sta 0,x
-                lda workword+1
-                sta 1,x
                 bra _semicolon_done
 
 _colonword:
+                ; ( nt )
+
                 ; if status bit 4 is still 1, we didn't compile any never-native
                 ; code so we can safely clear the NN flag
                 lda #%00010000
                 and status
                 beq +
-                ldy #1
-                lda (workword),y
+                lda (workword)
                 and #255-NN
-                sta (workword),y
+                sta (workword)
 +
-
-                ; CP is the byte that will be the address we use in the
-                ; header as the end-of-compile address (z_word). This is
-                ; six bytes down from the header
-                ldy #6
+                ; Calculate code size by subtracting xt from CP.
+                dex
+                dex
                 lda cp
-                sta (workword),y
-                iny
+                sta 0,x
                 lda cp+1
-                sta (workword),y
+                sta 1,x                 ; ( nt cp )
+
+                jsr w_swap              ; ( cp nt)
+                jsr w_name_to_int       ; ( cp xt )
+                jsr w_minus             ; ( cp-xt )
+
+                ; Use header status flags to calculate offset to code size
+                lda (workword)          ; Fetch status flags
+                and #DC+FP
+                lsr                     ; A=0 or 2 with FP in carry
+                adc #3
+                tay
+
+                lda 0,x                 ; LSB of code size
+                sta (workword),y        ; Write code size
+;TODO LC shuffle: if 1,x is nonzero then we have a large word, so shuffle or clamp based on NN
```

Comment:
still to implement: this is the edge case described in CREATE where we have a word that exceeds the single byte we reserved for the size.  if it's NN we can just mark it as 255 (since we won't be copying it), and if it's not we can MOVE it one byte along to write the size.

Reply:
Isn't that going to be an issue if the word has control structures that are larger than -128/+127 byte branches and actually use the JMP instruction?  You won't be able to search/replace those because words made with CREATE can also have arbitrary data in them.

We had discussed tracking whether, during compilation, a word had used any features that could potentially require a NN flag; the discussion was about not flagging user words NN if they were safe to natively compile.  If that was implemented, it might also be useful here.  I would think, however, that you'd need to start with the larger header and only shrink it at the end if it looks "safe" to do so.

Using that methodology, a word that uses `,`, `C,`, `[`, etc., for example, would not be moveable and would have to stick with the larger header (and also be flagged NN)

This would slow down compiling of small words (because they would get shuffled), but shouldn't affect execution speed.

Reply:
i thought a fair bit about this, and think the saving grace is that we only need an accurate length for words that are not NN.  

So in CREATE we initially reserve only a single byte length (which is going to be enough most of the time), and start speculatively compiling the word.  We are already tracking whether JMP gets compiled into the body one way or another (we have some tests for assembler literal cases too) so by the time we get to SEMICOLON here we know the correct value for NN.   

If we've compiled less than 256 bytes, we're fine already.  If we've compiled more than 256 bytes then either the word is relocatable, in which case we shuffle it up one byte and write the correct length, or it's NN in which case we don't need an accurate length value so can safely write 255 as a single byte placeholder and we're still fine.  It will look slightly misleading in SEE but not problematic; if we wanted we could even check for length=255 and NN=1 and show something like "size: 255+" for that rare case.

The main thing is that relocatable words bigger than 256 bytes should still work fine for applications that require it.

Reply:
I'll at least do the '+' tweak.  Re-generating the header if needed is also a good idea, though atm I assume it's a pretty rare occurrence so we can decide how annoying the not knowing is...

Reply:
For a quick "how bad is it" check, I compiled my code for accessing FAT32 on a compact flash card, figuring there would be some words of reasonable size in there.  I then asked for the sizes (speaking of this, the `words&sizes` code that is commented out in forth_code/user_words.fs will need to be updated or removed if we proceed, as it accesses the headers directly) and limited the list to only words that are 200 bytes and over, and I got the following list:
```
included 200 
cd 285 
string>filename 374 
read-line 401 
read-char 203 
open-file 482 
print-attrs 339 
walkdirectory 225 
fat.init 291 
finfo 529 
cf.read 314
```
Here's the code for some of these words so you can see what the complexity looks like (if you want the full source, it's available at [https://github.com/SamCoVT/ForthSoftware/blob/master/file.fs](https://github.com/SamCoVT/ForthSoftware/blob/master/file.fs)).  `print-attrs`, in particular, doesn't look that complex but is almost all native compiled assembly.  This is with stock settings for native compiling.
```
: cd ( caddr u -- ) ( Change the working directory )
   string>filename drop
   working_dir init_directory ( Start at the beginning of dir )
   ( Address of dirname in dirinfo format is on the stack )
   ['] findfile working_dir walkdirectory
   ( working_dir_fileid direntry )
   ?dup 0= if
      cr ." Can't find directory: " dup 0B type cr exit
   then
   ( working_dir_fileid direntry )
   swap drop ( Get rid of working_dir_fileid for now )
   dup >clustLOW @ swap >clustHI @ ( Get starting cluster )
   2dup d0= if ( Check for zero - meaning use root dir )
      ( Replace with correct cluster for root dir )
      2drop root_dir_first_cluster 2@
   then
   ( directory_firstcluster.d )
   ( Save everything to move to the new directory )
   working_dir >firstcluster 2! ( Fill in starting cluster )
   working_dir init_directory ; ( Fill in all the rest     )

( open-file - simplified for read-only access )
: open-file  ( c-addr u fam -- fileid ior )
   >r ( save fam )
   newfileid ( Allot memory for the file info )
   ( c-addr u fileid ) -rot 2 pick swap move ( filename )
   ( fileid ) r> over >attrib c! ( save fam )
   ( fileid ) working_dir  init_directory
   ( fileid ) ['] findfile working_dir walkdirectory
   ( fileid direntry|0 )
   ?dup 0= if
      cr ." Can't open file: " dup 0B type cr
      1 ( ior of 1 = Can't open file ) exit
   then
   ( fileid direntry ) swap >r ( save fileid )
   dup >clustLOW @ ( first cluster LSB )
   over >clustHI @ ( first cluster MSB )
   ( direntry firstcluster.d )
   2dup r@ >firstcluster 2! ( Save first cluster )
   2dup r@ >currentcluster 2! ( start at first cluster )
   ( direntry firstcluster.d )
   cluster>sector r@ >currentsector 2! ( starting sector )
   ( direntry ) dup >filesize 32bit@ r@ >filesize 2!
   ( direntry ) drop ( Done with direntry )
   r> ( bring back fileid structure )
   dup >fileposition d0!
   dup >linestart d0!
   0 ( no errors/exceptions )
;


: print-attrs ( direntry_addr -- ) ( Print file attributes )
   >attrib c@
   dup 01 and if [char] R emit else space then ( Read Only )
   dup 02 and if [char] H emit else space then ( Hidden    )
   dup 04 and if [char] S emit else space then ( System    )
   ( Skipping Volume ID )
   dup 10 and if [char] D emit else space then ( Directory )
       20 and if [char] A emit else space then ( Archive   )
   ;

: walkdirectory  ( xt fileid -- addr | 0)
   \ Run routine (given by xt) on every starting address of a
   \ directory entry in directory described by fileid.  The routine
   \ should be ( direntry_addr -- continueflag ) where continueflag
   \ is 1 if directory processing should continue. The return value
   \ of walkdirectory will either be the address of the directory
   \ entry it stopped on (in the sector buffer) or 0 if it reached
   \ the end.
   >r ( Save the fileid )
   begin
      \ Get the offset of the directory entry into the
      \ sectorbuffer.
      r@ directory_offset
      \ Turn it into an address in the sectorbuffer.
      sectorbuff +
      \ See if we have reached the end of the directory.
      dup dirend? if
         drop ( direntry ) r> 2drop ( xt and fileid)
         0 ( false ) exit then
      \ See if it's an unused entry.
      dup direntry_invalid? if
         \ Just put 1 on the stack to tell the routine to keep
         \ going.
         drop 1
      else
         \ It's a valid entry.  Call the processing routine on it.
         over execute
      then
      \ There should be a flag telling us if we should proceed to
      \ the next directory entry.
   while
         \ Proceed to the next entry
         r@ dir_next
   repeat
   \ If we made it here, we were told to stop.
   drop ( the xt )
   \ Leave the address of the directory entry on the stack.
   sectorbuff  r> directory_offset  +
;

: fat.init  ( -- )  ( Intialize the FAT values )
   0. sector>buffer \ get MBR
   \ Check the 4 partions for a FAT one.
   4 0 do
      sectorbuff >parttable \ Locate partition table
      i /partentry * + \ Pick a partition
      dup fat32part? if \ Check for fat32
         >partlba \ get starting LBA for this partition.
         32bit@   partition_lba_begin 2!
         ." Found FAT32 filesystem in partition " i .
         ."  at LBA " partition_lba_begin 2@ d.
      else
         drop
      then
   loop
   \ Make sure we found something.
   partition_lba_begin 2@ d0= if abort then
   \ Initialize all of the variables needed for operation.
   fat.init.vars
;

: finfo  ( fileid -- )  ( fileid info for debugging )
   cr ." FILEID: "          dup u.
   cr ." filename: "        dup >filename 0B type
   cr ." access mode: "     dup >attrib c@ u.
   cr ." first cluster: "   dup >firstcluster 2@ ud.
   cr ." current cluster: " dup >currentcluster 2@ ud.
   cr ." current sector: "  dup >currentsector 2@ ud.
   cr ." file position: "   dup >fileposition 2@ ud.
   cr ." file size: "       dup >filesize 2@ ud.
   cr ." linestart: "           >linestart 2@ ud. ;

: cf.read ( addr numsectors LBA.d -- )
   ( Read numsectors starting at LBA [double] to addr )
   \ Treat LBA double as two 16-bit cells.
   \ MSB is on top.  We can write this as a 16-bit cell to
   \ cf.lba+2, however we need to make sure we don't screw
   \ up the Drive and LBA bits (in the MSB)
   EFFF and ( make sure Drive is zero )
   E000 or  ( make sure LBA and reserved bits are 1 )
   cf.lba 2 + ! ( store high word of LBA )
   cf.lba !     ( store  low word of LBA )
   \ stack is now just ( addr numsectors )
   \ Tell the CF card how many sectors we want.
   dup cf.sectorcount c!
   
   \ Send the command to read (0x20)
   20 cf.status c!
   cf.busywait
   
   \ stack is still ( addr numsectors )
   \ Read all of the sectors.
   0 ?do
      \ Read one sector 0x200 or 512 bytes
      200 0 do
         \ Read one byte.
         cf.data c@ over c!
         \ Increment the address.
         1+
      loop
      cf.busywait \ Wait for next sector to be available.
   loop
   drop ( the address )
;
```

Reply:
Good point, updated `words&sizes` by defining 
```
: previousnt dup  2 + @  over  c@ 1 and 0= if
       $ff and over $ff00 and or 2dup > 0= if 256 - then
then nip ;
```